File: Leetcode/tree/preorderTraversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None
#        TreeNode{val: 1, left: None, right: TreeNode{val: 2, left: TreeNode{val: 3, left: None, right: None}, right: None}}


class Solution:
    def preorderTraversal(self, root: TreeNode) -> List[int]:
        if root:
            return [root.val] + Solution.preorderTraversal(self, root.left) + Solution.preorderTraversal(self, root.right)
        else:
            return []

File: Leetcode/preorderTraversal.java
import java.util.*;

class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode() {}
    TreeNode(int val) { this.val = val; }
    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

class Solution {
    ArrayList<Integer> arr;
    public List<Integer> preorderTraversal(TreeNode root) {
        arr = new ArrayList<>();
        dfs(root);

        return arr;
    }

    private void dfs(TreeNode node) {
        if(node == null) return;
        arr.add(node.val);
        dfs(node.left);
        dfs(node.right);
    }

    public static void main(String[] args) {
        Solution obj = new Solution();
    }
}